git_hg_sync/repo_synchronizer.py

```diff
@@ -47,14 +50,49 @@ def get_clone_repo(self) -> Repo:
     def _commit_has_mercurial_metadata(self, repo: Repo, git_commit: str) -> bool:
         return not all(char == "0" for char in self._git2hg(repo, git_commit))
 
-    def fetch_all_from_remote(self, repo: Repo, remote: str) -> None:
+    def fetch_all_from_remote(
+        self, repo: Repo, remote: str, verbose: bool = False
+    ) -> None:
         try:
-            repo.git.execute(["git", "-c", "cinnabar.graft=true", "fetch", remote])
-        except exc.GitCommandError as e:
+            self._log_git_execute(repo,
+                ["git", "-c", "cinnabar.graft=true", "fetch", "--tags", remote], verbose)
+
+        except GitCommandError as e:
             # can't fetch if repo is empty
             if "fatal: couldn't find remote ref HEAD" in e.stderr:
                 raise e
 
+        if remote.startswith("hg::"):
+            self._log_git_execute(repo,
+                ["git", "cinnabar", "fetch", "--tags"], verbose)
+
+    @staticmethod
+    def _log_git_execute(repo: Repo, command: list[str], verbose: bool = False) -> None:
+        proc = repo.git.execute(
+            command,
+            stdout_as_string=verbose,
+            as_process=verbose,
+        )
+        if verbose:
+            logger.info(f"Running `{' '.join(command)}` as PID {proc.pid}")
+
+            def stream_output(stream: io.BufferedReader, label: str) -> None:
+                for line in iter(stream.readline, b""):
+                    logger.info(f"{label}: {line.decode().strip()}")
```

Comment:
I think this could be rewritten more cleanly to reduce nesting (e.g., move the method definition outside the if statement).